File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function averageOfLevels(root: TreeNode | null): number[] {
  const levelSums = [];
  const levelCount = [];
  
  (function traverse(node: TreeNode, depth: number) {
    if (levelSums.length === depth) {
      levelSums.push(0);
      levelCount.push(0);
    }
    
    levelSums[depth] += node.val;
    levelCount[depth] += 1;
    
    node.left && traverse(node.left, depth + 1);
    node.right && traverse(node.right, depth + 1);
  })(root, 0);
  
  return levelSums.map((sum, i) => sum / levelCount[i]);
};

File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
    levelSums := make([]int, 0, 14)
    levelCount := make([]int, 0, 14)
    
    var traverse func (node *TreeNode, depth int)
    
    traverse = func (node *TreeNode, depth int) {
        if len(levelSums) == depth {
            levelSums = append(levelSums, 0.)
            levelCount = append(levelCount, 0)
        }
        
        levelSums[depth] += node.Val
        levelCount[depth] += 1
        
        if node.Left != nil {
            traverse(node.Left, depth + 1)
        }
        if node.Right != nil {
            traverse(node.Right, depth + 1)
        }
    }
    
    traverse(root, 0)
    
    avgs := make([]float64, len(levelSums))
    
    for i, sum := range levelSums {
        avgs[i] = float64(sum) / float64(levelCount[i])
    }
    
    return avgs
}